EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java

```diff
@@ -41,12 +47,36 @@ public PurchaseResponse placeOrder(Purchase purchase) {
         Set<OrderItem> orderItems = purchase.getOrderItems();
         orderItems.forEach(orderItem -> order.add(orderItem));
 
-        order.setShippingAddress(purchase.getShippingAddress());
-        order.setBillingAddress(purchase.getBillingAddress());
+        Address shippingAddress = purchase.getShippingAddress();
+        Address billingAddress = purchase.getBillingAddress();
+
+
+        Address existingShippingAddress = addressRepository.findAddressByCityAndCountryAndStateAndStreetAndZipCode(
+                shippingAddress.getCity(), shippingAddress.getCountry(), shippingAddress.getState(), shippingAddress.getStreet(), shippingAddress.getZipCode());
+
+        if (existingShippingAddress != null){
+            order.setShippingAddress(existingShippingAddress);
+        } else {
+            Address savedShippingAddress = addressRepository.save(shippingAddress);
+            order.setShippingAddress(savedShippingAddress);
+        }
+
+        Address existingBillingAddress = addressRepository.findAddressByCityAndCountryAndStateAndStreetAndZipCode(
+                billingAddress.getCity(), billingAddress.getCountry(), billingAddress.getState(), billingAddress.getStreet(), billingAddress.getZipCode());
+
+        if (existingBillingAddress != null){
+            order.setBillingAddress(existingBillingAddress);
+        } else {
+            Address savedBillingAddress = addressRepository.save(billingAddress);
+            order.setBillingAddress(savedBillingAddress);
+        }
+
+
         order.setStatus(OrderStatus.CREATED.toString());
 
 
         Customer customer = purchase.getCustomer();
+        System.out.println(customer);
```

Comment:
Why this println is needed? Assume this is some debug logging that supposed to be removed.
For proper logging dedicated libraries such as log4j should be used

Reply:
Removed